src/functions/arguments.ts

```diff
@@ -96,67 +97,163 @@ export function addMetaInfoFromArg(addDescr: AddFunctionDescription): FunctionDe
-    return (argPosition) => argPosition;
-  }
-  if (repeatingArg === 1) {
-    return (argPosition) => Math.min(argPosition, countArg);
-  }
-  const argBeforeRepeat = countArg - repeatingArg;
-  return (argPosition) => {
-    if (argPosition <= argBeforeRepeat) {
-      return argPosition;
+export function argTargeting(
+  functionDescription: FunctionDescription,
+  nbrArgSupplied: number
+): (argPosition: number) => number | undefined {
+  const valueIndexToArgPosition = new Map<number, number>();
+  const groupsOfRepeatingValues = functionDescription.nbrArgRepeating
+    ? Math.floor(
+        (nbrArgSupplied - functionDescription.minArgRequired) / functionDescription.nbrArgRepeating
+      )
+    : 0;
+  const nbrValueRepeating = functionDescription.nbrArgRepeating * groupsOfRepeatingValues;
+  const nbrValueOptional = nbrArgSupplied - functionDescription.minArgRequired - nbrValueRepeating;
+
+  let countValueSupplied = 0;
+  let countValueOptional = 0;
+
+  for (let i = 0; i < functionDescription.args.length; i++) {
+    const arg = functionDescription.args[i];
+
+    if (arg.optional || arg.default) {
+      if (countValueOptional < nbrValueOptional) {
+        valueIndexToArgPosition.set(countValueSupplied, i);
+        countValueSupplied++;
+      }
+      countValueOptional++;
+      continue;
     }
-    const argAfterRepeat = (argPosition - argBeforeRepeat) % repeatingArg || repeatingArg;
-    return argBeforeRepeat + argAfterRepeat;
+
+    if (arg.repeating) {
+      // As we know all repeating arguments are consecutive,
+      // --> we will treat all repeating arguments in one go
+      // --> the index i will be incremented by the number of repeating values at the end of the loop
+      for (let j = 0; j < groupsOfRepeatingValues; j++) {
+        for (let k = 0; k < functionDescription.nbrArgRepeating; k++) {
+          valueIndexToArgPosition.set(countValueSupplied, i + k);
+          countValueSupplied++;
+        }
+      }
+      i += functionDescription.nbrArgRepeating - 1;
+      continue;
+    }
+
+    // End case: it's a required argument
+    valueIndexToArgPosition.set(countValueSupplied, i);
+    countValueSupplied++;
+  }
+
+  return (argPosition: number) => {
+    return valueIndexToArgPosition.get(argPosition);
   };
 }
 
 //------------------------------------------------------------------------------
 // Argument validation
 //------------------------------------------------------------------------------
 
-export function validateArguments(args: ArgDefinition[]) {
-  let previousArgRepeating: boolean | undefined = false;
-  let previousArgOptional: boolean | undefined = false;
-  let previousArgDefault: boolean | undefined = false;
+export function validateArguments(
+  args: ArgDefinition[],
+  nbrArgOptional: number,
+  nbrArgRepeating: number
+) {
+  if (nbrArgRepeating && nbrArgOptional >= nbrArgRepeating) {
+    throw new Error(_t("Function ${name} has more optional arguments than repeatable ones."));
```

Comment:
wait, `${name}` is never replaced by any real value, is it ? Also, this error message doesn't need to be translated, it's for developers defining wrong arguments

Reply:
indeed :)